### ISSUE TITLE
Docs: kick start for the clueless

### ISSUE THREAD

**Issue by schwehr** [developmentseed/rio-stac#25]
I gave rio-stac a go, but as I've not used rasterio before, I am clearly missing something (yeah, I'm mostly stuck with legacy gdal's swig python code).

Here is what I tried on debian testing:

```
sudo apt-get install rasterio
virtualenv --system-site-packages --python=python3 ve
source ve/bin/activate
pip install --upgrade rasterio
pip install rio-stac

rasterio rio stac foo.tif 
rasterio stac foo.tif
```

Looking at https://readthedocs.org/projects/rasterio/downloads/pdf/latest/

```
4.19 Rio Plugins
Rio uses click-plugins to provide the ability to create additional subcommands using plugins developed outside
rasterio. This is ideal for commands that require additional dependencies beyond those used by rasterio, or that provide
functionality beyond the intended scope of rasterio.
For example, rio-mbtiles provides a command rio mbtiles to export a raster to an MBTiles file.
See click-plugins for more information on how to build these plugins in general.
To use these plugins with rio, add the commands to the rasterio.rio_plugins entry point in your setup.py file,
as described here and in rasterio/rio/main.py.
See the plugin registry for a list of available plugins
```

So trying this:

```
virtualenv --python=python3 ve3-r02
source ve3-r02/bin/activate
pip install --upgrade rio-stac
find . -name main.py | grep rasterio
# ./ve3-r02/lib/python3.9/site-packages/rasterio/rio/main.py

rio stac H13169_MB_1m_Ellipsoid_1of1.tif 
```

Does work...

```
{"type":"Feature","stac_version":"1.0.0","id":"H13169_MB_1m_Ellipsoid_1of1.tif","properties":{"proj:epsg":26917,"proj:geometry":{"type":"Polygon","coordinates":[[[393167.9999648061,2730943.0000648424],[393167.9999648061,2723229.9999351576],[407377.0000351939,2723229.9999351576],[407377.0000351939,2730943.0000648424],[393167.9999648061,2730943.0000648424]]]},"proj:bbox":[393167.9999648061,2723229.9999351576,407377.0000351939,2730943.0000648424],"proj:shape":[7712,14208],"proj:transform":[1.0000703878369819,0.0,393167.9999648061,0.0,-1.0001296848657761,2730943.0000648424,0.0,0.0,1.0],"datetime":"2021-11-08T17:28:34.720542Z"},"geometry":{"type":"Polygon","coordinates":[[[-82.05600859980615,24.690101662305853],[-82.05600859980615,24.619531295091686],[-81.9150632837156,24.619531295091686],[-81.9150632837156,24.690101662305853],[-82.05600859980615,24.690101662305853]]]},"links":[],"assets":{"asset":{"href":"/usr/local/google/home/schwehr/rio-stac/H13169_MB_1m_Ellipsoid_1of1.tif","raster:bands":[{"data_type":"float32","scale":1.0,"offset":0.0,"sampling":"area","nodata":"inf","statistics":{"mean":-32.33958251935924,"minimum":-36.26332092285156,"maximum":-23.28749656677246,"stddev":2.579397807199244,"valid_percent":54.8445228192446},"histogram":{"count":11,"min":-36.26332092285156,"max":-23.28749656677246,"buckets":[52223,58566,61887,63608,22481,16462,19766,14904,2356,1]}},{"data_type":"float32","scale":1.0,"offset":0.0,"sampling":"area","nodata":"inf","statistics":{"mean":0.05142540708045373,"minimum":1.0258333531965036e-05,"maximum":0.38972434401512146,"stddev":0.03645340210010217,"valid_percent":54.8445228192446},"histogram":{"count":11,"min":1.0258333531965036e-05,"max":0.38972434401512146,"buckets":[151352,105567,36301,12365,4321,1607,533,169,32,7]}}],"roles":[]}},"bbox":[-82.05600859980615,24.619531295091686,-81.9150632837156,24.690101662305853],"stac_extensions":["https://stac-extensions.github.io/projection/v1.0.0/schema.json","https://stac-extensions.github.io/raster/v1.1.0/schema.json"]}
```

```
pprint.pprint(d)
{'assets': {'asset': {'href': '/usr/local/google/home/schwehr/rio-stac/H13169_MB_1m_Ellipsoid_1of1.tif',
                      'raster:bands': [{'data_type': 'float32',
                                        'histogram': {'buckets': [52223,
                                                                  58566,
                                                                  61887,
                                                                  63608,
                                                                  22481,
                                                                  16462,
                                                                  19766,
                                                                  14904,
                                                                  2356,
                                                                  1],
                                                      'count': 11,
                                                      'max': -23.28749656677246,
                                                      'min': -36.26332092285156},
                                        'nodata': 'inf',
                                        'offset': 0.0,
                                        'sampling': 'area',
                                        'scale': 1.0,
                                        'statistics': {'maximum': -23.28749656677246,
                                                       'mean': -32.33958251935924,
                                                       'minimum': -36.26332092285156,
                                                       'stddev': 2.579397807199244,
                                                       'valid_percent': 54.8445228192446}},
                                       {'data_type': 'float32',
                                        'histogram': {'buckets': [151352,
                                                                  105567,
                                                                  36301,
                                                                  12365,
                                                                  4321,
                                                                  1607,
                                                                  533,
                                                                  169,
                                                                  32,
                                                                  7],
                                                      'count': 11,
                                                      'max': 0.38972434401512146,
                                                      'min': 1.0258333531965036e-05},
                                        'nodata': 'inf',
                                        'offset': 0.0,
                                        'sampling': 'area',
                                        'scale': 1.0,
                                        'statistics': {'maximum': 0.38972434401512146,
                                                       'mean': 0.05142540708045373,
                                                       'minimum': 1.0258333531965036e-05,
                                                       'stddev': 0.03645340210010217,
                                                       'valid_percent': 54.8445228192446}}],
                      'roles': []}},
 'bbox': [-82.05600859980615,
          24.619531295091686,
          -81.9150632837156,
          24.690101662305853],
 'geometry': {'coordinates': [[[-82.05600859980615, 24.690101662305853],
                               [-82.05600859980615, 24.619531295091686],
                               [-81.9150632837156, 24.619531295091686],
                               [-81.9150632837156, 24.690101662305853],
                               [-82.05600859980615, 24.690101662305853]]],
              'type': 'Polygon'},
 'id': 'H13169_MB_1m_Ellipsoid_1of1.tif',
 'links': [],
 'properties': {'datetime': '2021-11-08T17:28:34.720542Z',
                'proj:bbox': [393167.9999648061,
                              2723229.9999351576,
                              407377.0000351939,
                              2730943.0000648424],
                'proj:epsg': 26917,
                'proj:geometry': {'coordinates': [[[393167.9999648061,
                                                    2730943.0000648424],
                                                   [393167.9999648061,
                                                    2723229.9999351576],
                                                   [407377.0000351939,
                                                    2723229.9999351576],
                                                   [407377.0000351939,
                                                    2730943.0000648424],
                                                   [393167.9999648061,
                                                    2730943.0000648424]]],
                                  'type': 'Polygon'},
                'proj:shape': [7712, 14208],
                'proj:transform': [1.0000703878369819,
                                   0.0,
                                   393167.9999648061,
                                   0.0,
                                   -1.0001296848657761,
                                   2730943.0000648424,
                                   0.0,
                                   0.0,
                                   1.0]},
 'stac_extensions': ['https://stac-extensions.github.io/projection/v1.0.0/schema.json',
                     'https://stac-extensions.github.io/raster/v1.1.0/schema.json'],
 'stac_version': '1.0.0',
 'type': 'Feature'}
```

**Comment by vincentsarago** [comment]
@schwehr thanks for trying rio-stac. I'm not quite sure to understand what you didn't understand. The docs can be found over https://developmentseed.org/rio-stac/intro/ where we show how to use the `rio stac` command.

**Comment by schwehr** [comment]
If you follow what I assume is the default workflow for debian, you don't end up with a rio command.  So maybe being more explicit about requiring installing rasterio inside the virtual env so that rio and the plugin are actually installed.

**Comment by vincentsarago** [comment]
@schwehr 

I'm curious about `sudo apt-get install rasterio` I didn't know rasterio python module could be provided by `apt`

IMO, the default workflow for any python module should just be `pip install {module name}`.

> So maybe being more explicit about requiring installing rasterio inside the virtual env so that rio and the plugin are actually installed.

if you do `pip install rio-stac`, Pip will take care of the rasterio installation, so there is no need for other stuff here. 

```bash
$ virtualenv -p python3 ~/Dev/venv/clean   
created virtual environment CPython3.8.2.final.0-64 in 199ms

$ source clean/bin/activate

$ python --version
Python 3.8.2

$ venv which python
~/Dev/venv/clean/bin/python

# I'm using a Mac M1 so I cannot use rasterio wheel here
$ pip install rio-stac --no-binary rasterio 
...
Successfully built rio-stac rasterio
Installing collected packages: six, pyparsing, numpy, click, snuggs, python-dateutil, cligj, click-plugins, certifi, attrs, affine, rasterio, pystac, rio-stac
Successfully installed affine-2.3.0 attrs-21.2.0 certifi-2021.10.8 click-8.0.3 click-plugins-1.1.1 cligj-0.7.2 numpy-1.21.4 pyparsing-3.0.5 pystac-1.1.0 python-dateutil-2.8.2 rasterio-1.2.10 rio-stac-0.3.2 six-1.16.0 snuggs-1.4.7
...
$ which rio
/Users/vincentsarago/Dev/venv/clean/bin/rio

$ rio stac --help
Usage: rio stac [OPTIONS] INPUT

  Rasterio STAC plugin: Create a STAC Item for raster dataset.

Options:
  -d, --datetime TEXT             The date and time of the assets, in UTC (e.g
                                  2020-01-01, 2020-01-01T01:01:01).

  -e, --extension TEXT            STAC extension URL the Item implements.
  -c, --collection TEXT           The Collection ID that this item belongs to.
  --collection-url TEXT           Link to the STAC Collection.
  -p, --property NAME=VALUE       Additional property to add.
  --id TEXT                       Item id.
  -n, --asset-name TEXT           Asset name.
  --asset-href TEXT               Overwrite asset href.
  --asset-mediatype [COG|GEOJSON|GEOPACKAGE|GEOTIFF|HDF|HDF5|JPEG|JPEG2000|JSON|PNG|TEXT|TIFF|XML|auto]
                                  Asset media-type.
  --with-proj / --without-proj    Add the projection extension and properties
                                  (default to True).

  --with-raster / --without-raster
                                  Add the raster extension and properties
                                  (default to True).

  -o, --output PATH               Output file name
  --help                          Show this message and exit.
```

**Comment by schwehr** [comment]
I am curious why you would be surprised that apt does have rasterio :).  While I haven't used it much, I know it's pretty foundational for so much cool GIS work.

Folks on debian (or fink on mac, which I used to support, or many other systems), use package managers to install a much prebuilt and verified code.   Then --system-site-packages gets a big jump on installs.  So my only point of this issue (which I figured out while writing it) is that it would be helpful to have some text in the instructions about how to make sure the plugin is installed correctly for folks that are not already very familiar with rasterio plugins.  It looks like a normal user is not going to be able to write to the locations required to setup rasterio plugins for system packages.  They could with sudo, but I'd argue that users shouldn't pip install anything on the system python (after trying to unwedge so many people's machines after such adventures)

BTW, I finally gave up last year on packaging anything for Mac after 20 years of struggling with what Apple does with the command line tools, libs, interpreted languages, and compilers.  :(

And many thanks for writing rio-stac!!  It's cool.  I'm working on the Earth Engine catalog and comparing what I'm doing with what rio-stac does.

**Comment by kylebarron** [comment]
> If you follow what I assume is the default workflow for debian, you don't end up with a rio command.

Just pointing out that `rio` is not a command provided by `rio-stac`; it's a command provided by `rasterio` proper. If `apt install rasterio` doesn't provide a `rio` executable, then that would lead me to believe that the debian installation of `rasterio` is incorrect or incomplete.

Furthermore, is there a way using `apt` to install Python packages outside of a global environment? If not, seems that you'd never want to use `apt`-installed Python packages in production. I'd expect you'd hit the same issues as https://justinmayer.com/posts/homebrew-python-is-not-for-you/ with `apt` just like with homebrew.

FWIW, testing with the debian docker image, it looks like the `rio` command is (arguably incorrectly) found at `rasterio` instead. Seems like an "error" in the debian packaging of this package. `apt install rasterio python3-pip && pip install rio-stac` produces a `rasterio stac` entry point.

**Comment by vincentsarago** [comment]
Thanks everyone. @schwehr I understand the frustration but to me it seems that the documentation covers how to install and use rio-stac properly. We/I can't really cover all the way people will want to have their environment.

**Comment by schwehr** [comment]
No worries.  I got my local setup figured, so I'm good.  It's your call on the docs and this issue provides a place that folks might discover even if it is closed.

BTW...

I don't use debian images in production, but should think that a lot of folks would expect that system installed rasterios are correct.  I use google3 + blaze + build everything.  I only use debian for locally trying things out.

And I knew that homebrew wasn't for me after I made my own package manager for sunos 4, solaris, irix, vxworks, and linux that included TCL and Python builds that were loadable via NFS back in the 90s 🤪. I so should have written up that experience way back when.   It was cool and usable for a team of ~20 folks , but all the problems listed in [Homebrew Python Is Not For You](https://justinmayer.com/posts/homebrew-python-is-not-for-you/) have been around forever and keep getting rediscovered.
